pkg/virt-launcher/virtwrap/util/libvirt_helper.go

```diff
@@ -251,71 +254,150 @@ func StartLibvirt(stopChan chan struct{}) {
-					log.LogQemuLogLine(log.Log, scanner.Text())
-				}
+			if err := scanner.Err(); err != nil {
+				log.Log.Reason(err).Error("failed to read virtlogd logs")
+			}
+		}()
+
+		go func() {
+			defer close(exitChan)
+			_ = cmd.Wait()
+		}()
+
+		select {
+		case <-stopChan:
+			_ = cmd.Process.Kill()
+			return
+		case <-exitChan:
+			log.Log.Errorf("virtlogd exited, restarting")
+		}
 
-				if err := scanner.Err(); err != nil {
-					log.Log.Reason(err).Error("failed to read virtlogd logs")
+		// this sleep is to avoid consumming all resources in the
+		// event of a virtlogd crash loop.
+		time.Sleep(time.Second)
+	}
+}
+
+func startQEMUSeaBiosLogging(stopChan chan struct{}) {
+	const readSerialFilename = "readserial.py"
+	const readSerialScript = "/etc/qemu/scripts/" + readSerialFilename
+	const QEMUSeaBiosDebugPipeMode uint32 = 0666
+
+	err := syscall.Mkfifo(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("failed creating a pipe for sea bios debug logs: %s", err.Error()))
+		return
+	}
+
+	// Chmod is needed since umask is 0018. Therefore Mkfifo does not actually create a pipe with proper permissions.
+	err = syscall.Chmod(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("failed executing chmod on pipe for sea bios debug logs: %s", err.Error()))
+		return
+	}
+
+	for {
+		exitChan := make(chan struct{})
+		readSerialScriptCmd := exec.Command("/bin/python3", readSerialScript, "-nf", QEMUSeaBiosDebugPipe)
+
+		cmdStdout, err := readSerialScriptCmd.StdoutPipe()
+		if err != nil {
+			log.Log.Reason(err).Error(fmt.Sprintf("failed to get stdout pipe for %s. error: %s", readSerialFilename, err.Error()))
+			return
+		}
+
+		err = readSerialScriptCmd.Start()
+		if err != nil {
+			log.Log.Reason(err).Error(fmt.Sprintf("failed to start %s for QEMU SeaBios debug logging", readSerialFilename))
+			panic(err)
+		}
+
+		go func() {
+			var err error
+
+			// It can take a few seconds to the pipe to be created
+			for {
+				_, err = os.Stat(QEMUSeaBiosDebugPipe)
+				if !os.IsNotExist(err) {
+					break
 				}
-			}()
+				time.Sleep(time.Second)
+			}
 
-			go func() {
-				defer close(exitChan)
-				cmd.Wait()
-			}()
+			reader := bufio.NewReader(cmdStdout)
 
-			select {
-			case <-stopChan:
-				cmd.Process.Kill()
+			var line []byte
+			for line, _, err = reader.ReadLine(); err == nil; line, _, err = reader.ReadLine() {
+				log.LogQemuLogLine(log.Log, string(line))
+			}
+
+			if err != io.EOF {
+				log.Log.Reason(err).Error(fmt.Sprintf("%s failed with an error: %s", readSerialFilename, err.Error()))
 				return
-			case <-exitChan:
-				log.Log.Errorf("virtlogd exited, restarting")
 			}
 
-			// this sleep is to avoid consumming all resources in the
-			// event of a virtlogd crash loop.
-			time.Sleep(time.Second)
+		}()
+
+		go func() {
+			defer close(exitChan)
+			defer util.CloseIOAndCheckErr(cmdStdout, nil)
+
```

Comment:
I don't think you should explicitly close the pipe. I think it's being closed by the `wait()`

Reply:
Are you sure? I'm not sure that's right. I wait for "readSerialScriptCmd" which only consumes from the pipe by executing a python script. The pipe itself was created though Mkfifo above.

Reply:
Maybe I'm off here, but I thought it's:
your script  indeed consumes from the QEMUSeaBiosDebugPipe  pipe , and then you create another pipe cmdStdout for the script output (using StdoutPipe() ).   The cmdStdout  should be closed by the wait

Reply:
Maybe you wanted to close QEMUSeaBiosDebugPipe  ?